### ISSUE TITLE
Unsupported type List<dynamic>

### ISSUE THREAD

**Issue by Arpit1496** [jaclarke/__unofficial__edgedb-dart#35]
**Describe the bug**
When using query builders, the parsed Json does not provide type safety and throws runtime exceptions.

Unhandled exception:
type 'List<dynamic>' is not a subtype of type 'List<ReadCameraList_tags>'

In the above exception, Tags is a multi-linked type to base type Camera. The query does nested fetch of both.

**Reproduction**
///////////******* SCHEMA **********////////////
```esdl
type Camera {
required name: str;
multi tags: Tag;
}

type Tag {
required name:str;
}
```

///////********** QUERY ***********//////////

```edgeql
select Camera {
name,
tags {
   name,
 }
}
```

```dart
run : dart run build_runner build
fetch above generated query
```

**Expected behavior**
A clean fetch result should be obtained. But an exception is thrown stating that List<dynamic> is not a sub type of List<readCamera_Tags>

**Versions (please complete the following information):**
edgedb dart version: 0.4.0
edgedb: latest docker tag image (v4+)

![image](https://github.com/edgedb/edgedb-dart/assets/47178423/dc017a74-d65c-43a5-b4cf-c74088c973e6)

**Comment by Arpit1496** [comment]
When can we expect this to be fixed ? This is a breaking change. All our production queries are failing.

**Comment by Arpit1496** [comment]
Any update on this @raddevon ?

**Comment by jaclarke** [comment]
What version of dart are you using? The above schema/query seems to work for me using the latest version of dart:
```sh
$ dart --version
Dart SDK version: 3.2.5 (stable) (Tue Jan 16 15:02:13 2024 +0000) on "macos_arm64"
$ dart run lib/test_query.dart
[{"name":"test","tags":[{"name":"tag1"},{"name":"tag2"}]}]
```

<details>
<summary>`test_query.dart`</summary>

```dart
import 'dart:convert';

import 'package:edgedb/edgedb.dart';

import 'readCameraList.edgeql.dart';

void main() async {
  final client = createClient();

  final result = await client.readCameraList();

  print(jsonEncode(result));

  await client.close();
}
```
</details>

<details>
<summary>generated `readCameraList.edgeql.dart`</summary>

```dart
// AUTOGENERATED by 'edgeql_codegen' builder
// To re-generate use `dart run build_runner`

// ignore_for_file: no_leading_underscores_for_library_prefixes
import 'package:edgedb/src/codecs/codecs.dart' as _i1;
import 'package:edgedb/src/client.dart' as _i2;
import 'package:edgedb/src/primitives/types.dart' as _i3;

class ReadCameraList_tags {
  ReadCameraList_tags._fromMap(Map<String, dynamic> map) : name = map['name'];

  final String name;

  Map<String, dynamic> toMap() {
    return {'name': name};
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList {
  ReadCameraList._fromMap(Map<String, dynamic> map)
      : name = map['name'],
        tags = map['tags'];

  final String name;

  final List<ReadCameraList_tags> tags;

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'tags': tags,
    };
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

const _query = '''select Camera {
  name,
  tags: {
    name,
  }
}''';
final _outCodec = _i1.ObjectCodec(
  '8b2b0669066b549fa23886af970a54b0',
  [
    _i1.scalarCodecs['00000000000000000000000000000101']!,
    _i1.SetCodec<ReadCameraList_tags>(
      '36a088556139514991f853c718458e97',
      _i1.ObjectCodec(
        '9300f073485a533eb3eda5606d71a90a',
        [_i1.scalarCodecs['00000000000000000000000000000101']!],
        ['name'],
        [65],
        returnType: ReadCameraList_tags._fromMap,
      ),
    ),
  ],
  [
    'name',
    'tags',
  ],
  [
    65,
    109,
  ],
  returnType: ReadCameraList._fromMap,
);

extension ReadCameraListExtension on _i2.Executor {
  Future<List<ReadCameraList>> readCameraList() async {
    return await _i2.executeWithCodec<ReadCameraList>(
      this,
      'readCameraList',
      _outCodec,
      _i1.nullCodec,
      _i3.Cardinality.many,
      _query,
      null,
    );
  }
}

```
</details>

**Comment by Arpit1496** [comment]
Yes. Thanks for your time.

The issue has automatically been resolved in dart 3.4.2+.
I guess it was for dart version 3.1 only.

**Comment by Arpit1496** [comment]
This is my query: 
<details>
<summary> Query </summary>

```
select Camera {
  id,
  name,
  point: {
    id,
    lat,
    long,
  },
  tags: {
    id
  },
  uris: {
    id,
    cameraType,
    uri,
  },
  zones: {
    id,
    polygon,
    detectionsConfigs: {
      id,
      classLabels,
      objectConfidenceThresh,
      objectConfidenceChange,
      numberPlateAreaChange,
      numberPlateConfidenceMinThresh,
      numberPlateConfidenceThresh,
      numberPlateConfidenceChange,
      faceConfidenceMinThresh,
      faceConfidenceThresh,
      faceConfidenceChange,
      faceAreaChange,
      newObjectTimeoutSecs,
      directionPushStart,
      directionPushEnd,
      speedPushStart,
      speedPushEnd,
      stationaryTimeoutSecs,
      stationaryIntervalSecs,
      uploadType,
    },
    livePushInterval,
    aggPushInterval,
  },
  inferenceDevice,
  recordingDevice,
  frameDevice,
  eventDevice,
  segmentDevice,
}
```
</details>

Generated Code:
<details>
<summary> Generated Code </summary>

```
// AUTOGENERATED by 'edgeql_codegen' builder
// To re-generate use `dart run build_runner`

// ignore_for_file: no_leading_underscores_for_library_prefixes
import 'package:edgedb/src/codecs/codecs.dart' as _i1;
import 'package:edgedb/src/client.dart' as _i2;
import 'package:edgedb/src/primitives/types.dart' as _i3;

class ReadCameraList_point {
  ReadCameraList_point._fromMap(Map<String, dynamic> map)
      : id = map['id'],
        lat = map['lat'],
        long = map['long'];

  final String id;

  final double lat;

  final double long;

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'lat': lat,
      'long': long,
    };
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList_tags {
  ReadCameraList_tags._fromMap(Map<String, dynamic> map) : id = map['id'];

  final String id;

  Map<String, dynamic> toMap() {
    return {'id': id};
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList_uris {
  ReadCameraList_uris._fromMap(Map<String, dynamic> map)
      : id = map['id'],
        cameraType = map['cameraType'],
        uri = map['uri'];

  final String id;

  final String cameraType;

  final String uri;

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'cameraType': cameraType,
      'uri': uri,
    };
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList_zones_detectionsConfigs {
  ReadCameraList_zones_detectionsConfigs._fromMap(Map<String, dynamic> map)
      : id = map['id'],
        classLabels = map['classLabels'],
        objectConfidenceThresh = map['objectConfidenceThresh'],
        objectConfidenceChange = map['objectConfidenceChange'],
        numberPlateAreaChange = map['numberPlateAreaChange'],
        numberPlateConfidenceMinThresh = map['numberPlateConfidenceMinThresh'],
        numberPlateConfidenceThresh = map['numberPlateConfidenceThresh'],
        numberPlateConfidenceChange = map['numberPlateConfidenceChange'],
        faceConfidenceMinThresh = map['faceConfidenceMinThresh'],
        faceConfidenceThresh = map['faceConfidenceThresh'],
        faceConfidenceChange = map['faceConfidenceChange'],
        faceAreaChange = map['faceAreaChange'],
        newObjectTimeoutSecs = map['newObjectTimeoutSecs'],
        directionPushStart = map['directionPushStart'],
        directionPushEnd = map['directionPushEnd'],
        speedPushStart = map['speedPushStart'],
        speedPushEnd = map['speedPushEnd'],
        stationaryTimeoutSecs = map['stationaryTimeoutSecs'],
        stationaryIntervalSecs = map['stationaryIntervalSecs'],
        uploadType = map['uploadType'];

  final String id;

  final List<String> classLabels;

  final double objectConfidenceThresh;

  final double objectConfidenceChange;

  final double numberPlateAreaChange;

  final double numberPlateConfidenceMinThresh;

  final double numberPlateConfidenceThresh;

  final double numberPlateConfidenceChange;

  final double faceConfidenceMinThresh;

  final double faceConfidenceThresh;

  final double faceConfidenceChange;

  final double faceAreaChange;

  final double newObjectTimeoutSecs;

  final double directionPushStart;

  final double directionPushEnd;

  final double speedPushStart;

  final double speedPushEnd;

  final double stationaryTimeoutSecs;

  final double stationaryIntervalSecs;

  final String uploadType;

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'classLabels': classLabels,
      'objectConfidenceThresh': objectConfidenceThresh,
      'objectConfidenceChange': objectConfidenceChange,
      'numberPlateAreaChange': numberPlateAreaChange,
      'numberPlateConfidenceMinThresh': numberPlateConfidenceMinThresh,
      'numberPlateConfidenceThresh': numberPlateConfidenceThresh,
      'numberPlateConfidenceChange': numberPlateConfidenceChange,
      'faceConfidenceMinThresh': faceConfidenceMinThresh,
      'faceConfidenceThresh': faceConfidenceThresh,
      'faceConfidenceChange': faceConfidenceChange,
      'faceAreaChange': faceAreaChange,
      'newObjectTimeoutSecs': newObjectTimeoutSecs,
      'directionPushStart': directionPushStart,
      'directionPushEnd': directionPushEnd,
      'speedPushStart': speedPushStart,
      'speedPushEnd': speedPushEnd,
      'stationaryTimeoutSecs': stationaryTimeoutSecs,
      'stationaryIntervalSecs': stationaryIntervalSecs,
      'uploadType': uploadType,
    };
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList_zones {
  ReadCameraList_zones._fromMap(Map<String, dynamic> map)
      : id = map['id'],
        polygon = map['polygon'],
        detectionsConfigs = map['detectionsConfigs'],
        livePushInterval = map['livePushInterval'],
        aggPushInterval = map['aggPushInterval'];

  final String id;

  final List<double> polygon;

  final List<ReadCameraList_zones_detectionsConfigs> detectionsConfigs;

  final double livePushInterval;

  final double aggPushInterval;

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'polygon': polygon,
      'detectionsConfigs': detectionsConfigs,
      'livePushInterval': livePushInterval,
      'aggPushInterval': aggPushInterval,
    };
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList_inferenceDevice {
  ReadCameraList_inferenceDevice._fromMap(Map<String, dynamic> map)
      : id = map['id'];

  final String id;

  Map<String, dynamic> toMap() {
    return {'id': id};
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList_recordingDevice {
  ReadCameraList_recordingDevice._fromMap(Map<String, dynamic> map)
      : id = map['id'];

  final String id;

  Map<String, dynamic> toMap() {
    return {'id': id};
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList_frameDevice {
  ReadCameraList_frameDevice._fromMap(Map<String, dynamic> map)
      : id = map['id'];

  final String id;

  Map<String, dynamic> toMap() {
    return {'id': id};
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList_eventDevice {
  ReadCameraList_eventDevice._fromMap(Map<String, dynamic> map)
      : id = map['id'];

  final String id;

  Map<String, dynamic> toMap() {
    return {'id': id};
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList_segmentDevice {
  ReadCameraList_segmentDevice._fromMap(Map<String, dynamic> map)
      : id = map['id'];

  final String id;

  Map<String, dynamic> toMap() {
    return {'id': id};
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList {
  ReadCameraList._fromMap(Map<String, dynamic> map)
      : id = map['id'],
        name = map['name'],
        point = map['point'],
        tags = map['tags'],
        uris = map['uris'],
        zones = map['zones'],
        inferenceDevice = map['inferenceDevice'],
        recordingDevice = map['recordingDevice'],
        frameDevice = map['frameDevice'],
        eventDevice = map['eventDevice'],
        segmentDevice = map['segmentDevice'];

  final String id;

  final String name;

  final ReadCameraList_point? point;

  final List<ReadCameraList_tags> tags;

  final List<ReadCameraList_uris> uris;

  final List<ReadCameraList_zones> zones;

  final ReadCameraList_inferenceDevice? inferenceDevice;

  final ReadCameraList_recordingDevice? recordingDevice;

  final ReadCameraList_frameDevice? frameDevice;

  final ReadCameraList_eventDevice? eventDevice;

  final ReadCameraList_segmentDevice? segmentDevice;

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'name': name,
      'point': point,
      'tags': tags,
      'uris': uris,
      'zones': zones,
      'inferenceDevice': inferenceDevice,
      'recordingDevice': recordingDevice,
      'frameDevice': frameDevice,
      'eventDevice': eventDevice,
      'segmentDevice': segmentDevice,
    };
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

const _query = '''with cursor := <optional uuid>\$cursor
select Camera {
  id,
  name,
  point: {
    id,
    lat,
    long,
  },
  tags: {
    id
  },
  uris: {
    id,
    cameraType,
    uri,
  },
  zones: {
    id,
    polygon,
    detectionsConfigs: {
      id,
      classLabels,
      objectConfidenceThresh,
      objectConfidenceChange,
      numberPlateAreaChange,
      numberPlateConfidenceMinThresh,
      numberPlateConfidenceThresh,
      numberPlateConfidenceChange,
      faceConfidenceMinThresh,
      faceConfidenceThresh,
      faceConfidenceChange,
      faceAreaChange,
      newObjectTimeoutSecs,
      directionPushStart,
      directionPushEnd,
      speedPushStart,
      speedPushEnd,
      stationaryTimeoutSecs,
      stationaryIntervalSecs,
      uploadType,
    },
    livePushInterval,
    aggPushInterval,
  },
  inferenceDevice,
  recordingDevice,
  frameDevice,
  eventDevice,
  segmentDevice,
} filter any({
  all({
    contains(<optional array<uuid>>\$idIds, .id),
    contains(str_lower(.name), str_lower(<optional str>\$nameContains)),
    any(.point.id in array_unpack(<optional array<uuid>>\$pointIds)) if len(<optional array<uuid>>\$pointIds) ?= 0 else {},
    any(.tags.id in array_unpack(<optional array<uuid>>\$tagsIds)) if len(<optional array<uuid>>\$tagsIds) ?= 0 else {},
    any(.uris.id in array_unpack(<optional array<uuid>>\$urisIds)) if len(<optional array<uuid>>\$urisIds) ?= 0 else {},
    any(.zones.id in array_unpack(<optional array<uuid>>\$zonesIds)) if len(<optional array<uuid>>\$zonesIds) ?= 0 else {},
    any(.inferenceDevice.id in array_unpack(<optional array<uuid>>\$inferenceDeviceIds)) if len(<optional array<uuid>>\$inferenceDeviceIds) ?= 0 else {},
    any(.recordingDevice.id in array_unpack(<optional array<uuid>>\$recordingDeviceIds)) if len(<optional array<uuid>>\$recordingDeviceIds) ?= 0 else {},
    any(.frameDevice.id in array_unpack(<optional array<uuid>>\$frameDeviceIds)) if len(<optional array<uuid>>\$frameDeviceIds) ?= 0 else {},
    any(.eventDevice.id in array_unpack(<optional array<uuid>>\$eventDeviceIds)) if len(<optional array<uuid>>\$eventDeviceIds) ?= 0 else {},
    any(.segmentDevice.id in array_unpack(<optional array<uuid>>\$segmentDeviceIds)) if len(<optional array<uuid>>\$segmentDeviceIds) ?= 0 else {},
    .id > cursor,
  }),
  all({
    contains(<optional array<uuid>>\$idIds, .id),
    contains(str_lower(.name), str_lower(<optional str>\$nameContains)),
    str_lower(.name) > str_lower(<optional str>\$minName),
    str_lower(.name) < str_lower(<optional str>\$maxName),
    any(.point.id in array_unpack(<optional array<uuid>>\$pointIds)),
    any(.tags.id in array_unpack(<optional array<uuid>>\$tagsIds)),
    any(.uris.id in array_unpack(<optional array<uuid>>\$urisIds)),
    any(.zones.id in array_unpack(<optional array<uuid>>\$zonesIds)),
    any(.inferenceDevice.id in array_unpack(<optional array<uuid>>\$inferenceDeviceIds)),
    any(.recordingDevice.id in array_unpack(<optional array<uuid>>\$recordingDeviceIds)),
    any(.frameDevice.id in array_unpack(<optional array<uuid>>\$frameDeviceIds)),
    any(.eventDevice.id in array_unpack(<optional array<uuid>>\$eventDeviceIds)),
    any(.segmentDevice.id in array_unpack(<optional array<uuid>>\$segmentDeviceIds)),
  })
})
order by
.name if <optional str>\$sortBy = \'nameAsc\' else <str>{} empty last
then .name if <optional str>\$sortBy = \'nameDesc\' else <str>{} DESC empty last
limit <int32>\$limit''';
final _outCodec = _i1.ObjectCodec(
  '08cddbdd2beb55a0ad72ada24ccff3f7',
  [
    _i1.scalarCodecs['00000000000000000000000000000100']!,
    _i1.scalarCodecs['00000000000000000000000000000101']!,
    _i1.ObjectCodec(
      'a68b087cfc7857d4837666fd57ca901d',
      [
        _i1.scalarCodecs['00000000000000000000000000000100']!,
        _i1.scalarCodecs['00000000000000000000000000000106']!,
        _i1.scalarCodecs['00000000000000000000000000000106']!,
      ],
      [
        'id',
        'lat',
        'long',
      ],
      [
        65,
        65,
        65,
      ],
      returnType: ReadCameraList_point._fromMap,
    ),
    _i1.SetCodec<ReadCameraList_tags>(
      'e3209afecdf25e41afac4c66ccfd7050',
      _i1.ObjectCodec(
        '99766c5f0f4c52c0bcd221f51697aa5c',
        [_i1.scalarCodecs['00000000000000000000000000000100']!],
        ['id'],
        [65],
        returnType: ReadCameraList_tags._fromMap,
      ),
    ),
    _i1.SetCodec<ReadCameraList_uris>(
      'ad683beacf1a50588696b4a6307e4e96',
      _i1.ObjectCodec(
        'bac613db46bf5213925f84a8bbcc6001',
        [
          _i1.scalarCodecs['00000000000000000000000000000100']!,
          _i1.EnumCodec('ff3e6b4eb3d611ee8d3ecbc4f351ab3f'),
          _i1.scalarCodecs['00000000000000000000000000000101']!,
        ],
        [
          'id',
          'cameraType',
          'uri',
        ],
        [
          65,
          65,
          65,
        ],
        returnType: ReadCameraList_uris._fromMap,
      ),
    ),
    _i1.SetCodec<ReadCameraList_zones>(
      'e10d95dbb1f35ab886b7586338b73e12',
      _i1.ObjectCodec(
        'e598268abc8c5689acf009d5a0adb138',
        [
          _i1.scalarCodecs['00000000000000000000000000000100']!,
          _i1.ArrayCodec<double>(
            '6eccf9e113d857c8b894b0fea4fe2f46',
            'array<std|float32>',
            _i1.scalarCodecs['00000000000000000000000000000106']!,
            -1,
          ),
          _i1.SetCodec<ReadCameraList_zones_detectionsConfigs>(
            'bef95cf94b355ea9a18a5663af419a9b',
            _i1.ObjectCodec(
              '5ef7779a366450f59eeadb0637722b2b',
              [
                _i1.scalarCodecs['00000000000000000000000000000100']!,
                _i1.SetCodec<String>(
                  'e3971a2695ec552290a334ce6486ec0e',
                  _i1.EnumCodec('fea744ebb3d611ee843d4b993ae91a59'),
                ),
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.scalarCodecs['00000000000000000000000000000106']!,
                _i1.EnumCodec('fea75536b3d611ee801bcbc1856dbf3f'),
              ],
              [
                'id',
                'classLabels',
                'objectConfidenceThresh',
                'objectConfidenceChange',
                'numberPlateAreaChange',
                'numberPlateConfidenceMinThresh',
                'numberPlateConfidenceThresh',
                'numberPlateConfidenceChange',
                'faceConfidenceMinThresh',
                'faceConfidenceThresh',
                'faceConfidenceChange',
                'faceAreaChange',
                'newObjectTimeoutSecs',
                'directionPushStart',
                'directionPushEnd',
                'speedPushStart',
                'speedPushEnd',
                'stationaryTimeoutSecs',
                'stationaryIntervalSecs',
                'uploadType',
              ],
              [
                65,
                109,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
                65,
              ],
              returnType: ReadCameraList_zones_detectionsConfigs._fromMap,
            ),
          ),
          _i1.scalarCodecs['00000000000000000000000000000106']!,
          _i1.scalarCodecs['00000000000000000000000000000106']!,
        ],
        [
          'id',
          'polygon',
          'detectionsConfigs',
          'livePushInterval',
          'aggPushInterval',
        ],
        [
          65,
          65,
          109,
          65,
          65,
        ],
        returnType: ReadCameraList_zones._fromMap,
      ),
    ),
    _i1.ObjectCodec(
      '77f4e886c53b58049f012fed3cbe54fe',
      [_i1.scalarCodecs['00000000000000000000000000000100']!],
      ['id'],
      [65],
      returnType: ReadCameraList_inferenceDevice._fromMap,
    ),
    _i1.ObjectCodec(
      '77f4e886c53b58049f012fed3cbe54fe',
      [_i1.scalarCodecs['00000000000000000000000000000100']!],
      ['id'],
      [65],
      returnType: ReadCameraList_recordingDevice._fromMap,
    ),
    _i1.ObjectCodec(
      '77f4e886c53b58049f012fed3cbe54fe',
      [_i1.scalarCodecs['00000000000000000000000000000100']!],
      ['id'],
      [65],
      returnType: ReadCameraList_frameDevice._fromMap,
    ),
    _i1.ObjectCodec(
      '77f4e886c53b58049f012fed3cbe54fe',
      [_i1.scalarCodecs['00000000000000000000000000000100']!],
      ['id'],
      [65],
      returnType: ReadCameraList_eventDevice._fromMap,
    ),
    _i1.ObjectCodec(
      '77f4e886c53b58049f012fed3cbe54fe',
      [_i1.scalarCodecs['00000000000000000000000000000100']!],
      ['id'],
      [65],
      returnType: ReadCameraList_segmentDevice._fromMap,
    ),
  ],
  [
    'id',
    'name',
    'point',
    'tags',
    'uris',
    'zones',
    'inferenceDevice',
    'recordingDevice',
    'frameDevice',
    'eventDevice',
    'segmentDevice',
  ],
  [
    65,
    65,
    111,
    109,
    109,
    109,
    111,
    111,
    111,
    111,
    111,
  ],
  returnType: ReadCameraList._fromMap,
);
final _inCodec = _i1.ObjectCodec(
  'c85d925ada275489baa52bb4990618f5',
  [
    _i1.scalarCodecs['00000000000000000000000000000100']!,
    _i1.scalarCodecs['00000000000000000000000000000104']!,
    _i1.ArrayCodec<String>(
      'a1b4292529ed5dbeb3a0d773d9443ba8',
      'array<std|uuid>',
      _i1.scalarCodecs['00000000000000000000000000000100']!,
      -1,
    ),
    _i1.scalarCodecs['00000000000000000000000000000101']!,
    _i1.ArrayCodec<String>(
      'a1b4292529ed5dbeb3a0d773d9443ba8',
      'array<std|uuid>',
      _i1.scalarCodecs['00000000000000000000000000000100']!,
      -1,
    ),
    _i1.ArrayCodec<String>(
      'a1b4292529ed5dbeb3a0d773d9443ba8',
      'array<std|uuid>',
      _i1.scalarCodecs['00000000000000000000000000000100']!,
      -1,
    ),
    _i1.ArrayCodec<String>(
      'a1b4292529ed5dbeb3a0d773d9443ba8',
      'array<std|uuid>',
      _i1.scalarCodecs['00000000000000000000000000000100']!,
      -1,
    ),
    _i1.ArrayCodec<String>(
      'a1b4292529ed5dbeb3a0d773d9443ba8',
      'array<std|uuid>',
      _i1.scalarCodecs['00000000000000000000000000000100']!,
      -1,
    ),
    _i1.ArrayCodec<String>(
      'a1b4292529ed5dbeb3a0d773d9443ba8',
      'array<std|uuid>',
      _i1.scalarCodecs['00000000000000000000000000000100']!,
      -1,
    ),
    _i1.ArrayCodec<String>(
      'a1b4292529ed5dbeb3a0d773d9443ba8',
      'array<std|uuid>',
      _i1.scalarCodecs['00000000000000000000000000000100']!,
      -1,
    ),
    _i1.ArrayCodec<String>(
      'a1b4292529ed5dbeb3a0d773d9443ba8',
      'array<std|uuid>',
      _i1.scalarCodecs['00000000000000000000000000000100']!,
      -1,
    ),
    _i1.ArrayCodec<String>(
      'a1b4292529ed5dbeb3a0d773d9443ba8',
      'array<std|uuid>',
      _i1.scalarCodecs['00000000000000000000000000000100']!,
      -1,
    ),
    _i1.ArrayCodec<String>(
      'a1b4292529ed5dbeb3a0d773d9443ba8',
      'array<std|uuid>',
      _i1.scalarCodecs['00000000000000000000000000000100']!,
      -1,
    ),
    _i1.scalarCodecs['00000000000000000000000000000101']!,
    _i1.scalarCodecs['00000000000000000000000000000101']!,
    _i1.scalarCodecs['00000000000000000000000000000101']!,
  ],
  [
    'cursor',
    'limit',
    'idIds',
    'nameContains',
    'pointIds',
    'tagsIds',
    'urisIds',
    'zonesIds',
    'inferenceDeviceIds',
    'recordingDeviceIds',
    'frameDeviceIds',
    'eventDeviceIds',
    'segmentDeviceIds',
    'minName',
    'maxName',
    'sortBy',
  ],
  [
    111,
    65,
    111,
    111,
    111,
    111,
    111,
    111,
    111,
    111,
    111,
    111,
    111,
    111,
    111,
    111,
  ],
);

extension ReadCameraListExtension on _i2.Executor {
  Future<List<ReadCameraList>> readCameraList({
    String? cursor,
    required int limit,
    List<String>? idIds,
    String? nameContains,
    List<String>? pointIds,
    List<String>? tagsIds,
    List<String>? urisIds,
    List<String>? zonesIds,
    List<String>? inferenceDeviceIds,
    List<String>? recordingDeviceIds,
    List<String>? frameDeviceIds,
    List<String>? eventDeviceIds,
    List<String>? segmentDeviceIds,
    String? minName,
    String? maxName,
    String? sortBy,
  }) async {
    return await _i2.executeWithCodec<ReadCameraList>(
      this,
      'readCameraList',
      _outCodec,
      _inCodec,
      _i3.Cardinality.many,
      _query,
      <String, dynamic>{
        'cursor': cursor,
        'limit': limit,
        'idIds': idIds,
        'nameContains': nameContains,
        'pointIds': pointIds,
        'tagsIds': tagsIds,
        'urisIds': urisIds,
        'zonesIds': zonesIds,
        'inferenceDeviceIds': inferenceDeviceIds,
        'recordingDeviceIds': recordingDeviceIds,
        'frameDeviceIds': frameDeviceIds,
        'eventDeviceIds': eventDeviceIds,
        'segmentDeviceIds': segmentDeviceIds,
        'minName': minName,
        'maxName': maxName,
        'sortBy': sortBy,
      },
    );
  }
}

```
</details>

This fails with :
```
Unhandled exception:
type 'List<dynamic>' is not a subtype of type 'List<ReadCameraList_tags>'
#0      BaseProtocol.execute (package:edgedb/src/base_proto.dart:707:7)
<asynchronous suspension>
#1      BaseProtocol.fetch (package:edgedb/src/base_proto.dart:414:7)
<asynchronous suspension>
#2      ClientConnectionHolder._retryingFetch (package:edgedb/src/client.dart:157:18)
<asynchronous suspension>
#3      Client._executeWithCodec (package:edgedb/src/client.dart:757:14)
<asynchronous suspension>
#4      ReadCameraListExtension.readCameraList (package:astra_server/src/services/camera/camera/edgeql/readCameraList.edgeql.dart:760:12)
<asynchronous suspension>
#5      main.<anonymous closure> (package:astra_server/src/services/camera/camera/service.dart:221:38)
<asynchronous suspension>
```

**Comment by Arpit1496** [comment]
Dart version: 3.2.5(stable)
edgedb: 0.4.0

<details>
<summary> Pubsec.yaml </summary>

dependencies:
  astra_common:
    path: ../astra_common
  bcrypt: ^1.1.3
  edgedb: ^0.4.0
  grpc: ^3.2.3
  protobuf: ^3.1.0

dev_dependencies:
  build_runner: ^2.4.6
  hotreloader: ^4.1.0
  lint: ^2.0.0
  protoc_builder: ^0.4.0

</details>

Now I am getting static compile time error. 
![image](https://github.com/edgedb/edgedb-dart/assets/47178423/43cdb1b0-c6cd-4ec9-9b15-a7983dba76bd)

I have done dart update with flutter clean and re built all generated code.

**Comment by Arpit1496** [comment]
I have done same thing on my system (after using flutter clean). It still fails. 

Dart SDK version: 3.2.5 (stable) (Tue Jan 16 15:02:13 2024 +0000) on "linux_x64"

I have used the above query and this is my generated file:

```
select Camera {
  name,
  tags: {
    name,
  },
}
```

<details>
<summary>Generated Code</summary>

```
// AUTOGENERATED by 'edgeql_codegen' builder
// To re-generate use `dart run build_runner`

// ignore_for_file: no_leading_underscores_for_library_prefixes
import 'package:edgedb/src/codecs/codecs.dart' as _i1;
import 'package:edgedb/src/client.dart' as _i2;
import 'package:edgedb/src/primitives/types.dart' as _i3;

class ReadCameraList_tags {
  ReadCameraList_tags._fromMap(Map<String, dynamic> map) : name = map['name'];

  final String name;

  Map<String, dynamic> toMap() {
    return {'name': name};
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

class ReadCameraList {
  ReadCameraList._fromMap(Map<String, dynamic> map)
      : name = map['name'],
        tags = map['tags'];

  final String name;

  final List<ReadCameraList_tags> tags;

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'tags': tags,
    };
  }

  Map<String, dynamic> toJson() {
    return toMap();
  }
}

const _query = '''select Camera {
  name,
  tags: {
    name,
  },
}''';
final _outCodec = _i1.ObjectCodec(
  '8b2b0669066b549fa23886af970a54b0',
  [
    _i1.scalarCodecs['00000000000000000000000000000101']!,
    _i1.SetCodec<ReadCameraList_tags>(
      '36a088556139514991f853c718458e97',
      _i1.ObjectCodec(
        '9300f073485a533eb3eda5606d71a90a',
        [_i1.scalarCodecs['00000000000000000000000000000101']!],
        ['name'],
        [65],
        returnType: ReadCameraList_tags._fromMap,
      ),
    ),
  ],
  [
    'name',
    'tags',
  ],
  [
    65,
    109,
  ],
  returnType: ReadCameraList._fromMap,
);

extension ReadCameraListExtension on _i2.Executor {
  Future<List<ReadCameraList>> readCameraList() async {
    return await _i2.executeWithCodec<ReadCameraList>(
      this,
      'readCameraList',
      _outCodec,
      _i1.nullCodec,
      _i3.Cardinality.many,
      _query,
      null,
    );
  }
}
```

</details>

And it Fails with same error:

![image](https://github.com/edgedb/edgedb-dart/assets/47178423/386968cf-dd69-4a4a-bfbb-30424caca663)

**Comment by Arpit1496** [comment]
@raddevon got the bug.

It was in casting "empty" lists during runtime. Another compile time bugs  were for EnumCodex where Nullable positional parameter was not passed.

I have fixed both the bugs in this pr:

https://github.com/edgedb/edgedb-dart/pull/36

Kindly find the time and have a look as this may be breaking changes for all dart users.

**Comment by jaclarke** [comment]
Thanks for the PR @Arpit1496, and sorry it's taken a while for me to get back around to looking at this issue.

It looks like the `List<dynamic> is not a subtype of ...` issue is due to the set codec following a slightly different codepath for empty sets and sets of arrays, which doesn't properly cast the type of the returned `List` with the specified type parameter. And the enum codegen bug is regression from https://github.com/edgedb/edgedb-dart/pull/29, which it seems we didn't catch.

I'm going to close #36 in favour of this PR: #37, which fixes the root issue with the casting bug, and also adds some tests.

**Comment by jaclarke** [comment]
Fix has been published in v0.4.1
